src/libraries/System.Net.Sockets/src/System/Net/Connections/SocketsConnectionFactory.cs

```diff
@@ -0,0 +1,163 @@
+        /// </summary>
+        /// <param name="addressFamily">The <see cref="AddressFamily"/> to forward to the socket.</param>
+        /// <param name="socketType">The <see cref="SocketType"/> to forward to the socket.</param>
+        /// <param name="protocolType">The <see cref="ProtocolType"/> to forward to the socket.</param>
+        public SocketsConnectionFactory(
+            AddressFamily addressFamily,
+            SocketType socketType,
+            ProtocolType protocolType)
+        {
+            _addressFamily = addressFamily;
+            _socketType = socketType;
+            _protocolType = protocolType;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SocketsConnectionFactory"/> class
+        /// that will forward <see cref="AddressFamily.InterNetworkV6"/> to the Socket constructor.
+        /// </summary>
+        /// <param name="socketType">The <see cref="SocketType"/> to forward to the socket.</param>
+        /// <param name="protocolType">The <see cref="ProtocolType"/> to forward to the socket.</param>
+        /// <remarks>The created socket will be an IPv6 socket with <see cref="Socket.DualMode"/> enabled.</remarks>
+        public SocketsConnectionFactory(SocketType socketType, ProtocolType protocolType)
+            : this(AddressFamily.InterNetworkV6, socketType, protocolType)
+        {
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask<Connection> ConnectAsync(
+            EndPoint? endPoint,
+            IConnectionProperties? options = null,
+            CancellationToken cancellationToken = default)
+        {
+            if (endPoint == null) throw new ArgumentNullException(nameof(endPoint));
+            if (cancellationToken.IsCancellationRequested)
+            {
+                throw new TaskCanceledException(s_cancellationMessage, innerException: null, cancellationToken);
+            }
+
+            Socket socket = CreateSocket(_addressFamily, _socketType, _protocolType, endPoint, options);
+
+            try
+            {
+                using var args = new TaskSocketAsyncEventArgs();
+                args.RemoteEndPoint = endPoint;
+
+                if (socket.ConnectAsync(args))
+                {
+                    using (cancellationToken.UnsafeRegister(o => Socket.CancelConnectAsync((SocketAsyncEventArgs)o!), args))
+                    {
+                        await args.Task.ConfigureAwait(false);
+                    }
+                }
+
+                if (args.SocketError != SocketError.Success)
+                {
+                    SocketException ex = new SocketException((int)args.SocketError);
+                    if (args.SocketError == SocketError.OperationAborted && cancellationToken.IsCancellationRequested)
+                    {
+                        throw new TaskCanceledException(s_cancellationMessage, ex, cancellationToken);
```

Comment:
We can use CancellationToken.ThrowIfCancellationRequested here.

Reply:
It will throw `OperationAbortedException` we probably want `TaskCanceledException`, don't we?

Reply:
I don't think it matters. We use ThrowIfCancellationRequested in lots of places. I think TaskCanceledException is only for compatibility with existing APIs that already threw it.

Reply:
Will change then, but would be nice to have a guideline point on this, if there isn't already.

Reply:
We should never have shipped TaskCanceledException. It's one of the places I learned a good lesson about being over aggressive introducing new exception types. It's purpose was to be an OCE that also let you know what specific Task instance was canceled, but in practice it almost never contains that task, and even if it did, knowing which task was canceled is almost always useless. But as happens devs started adding catch blocks for TCE rather than OCE. Face palm. At this point, the only reason to throw TCE is if there are compat concerns with such catch blocks, and devs should not be writing new such catch blocks.

Reply:
I would make the `InnerException` be a `NetworkErrorHelper.MapSocketException(ex)`.

Reply:
I don't think we even need an inner exception here. The cancellationToken fired, so you get an OperationCancelledException. The fact that there's technically a SocketException with SocketErrorCode == OperationCanceled under the covers is really just an implementation detail.

Reply:
> I don't think we even need an inner exception here. The cancellationToken fired, so you get an OperationCancelledException. The fact that there's technically a SocketException with SocketErrorCode == OperationCanceled under the covers is really just an implementation detail.

It looks like `NetworkStream.WriteAsync` is not filling in an `InnerException` so this seems fine.